src/cata_ui.cpp

```diff
@@ -0,0 +1,1038 @@
+{
+}
+
+void ui_vertical_list::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    unsigned int start_line = window_scroll;
+    unsigned int end_line = start_line + get_rect().size_y;
+
+    unsigned int available_space = get_rect().size_x - 2; // 2 for scroll bar and spacer
+
+    for( unsigned int line = start_line; line < end_line && line < text.size(); line++ ) {
+        auto txt = text[line];
+        if( txt.size() > available_space ) {
+            txt = txt.substr( 0, available_space );
+        }
+        if( scroll == line ) {
+            mvwprintz( win, get_ay() + line - start_line, get_ax() + 2, hilite(text_color), txt.c_str() );
+        } else {
+            mvwprintz( win, get_ay() + line - start_line, get_ax() + 2, text_color, txt.c_str() );
+        }
+    }
+
+    draw_scrollbar( win, scroll, get_rect().size_y, text.size(), get_ay(), get_ax(), bar_color, false );
+}
+
+void ui_vertical_list::set_text( std::vector<std::string> text )
+{
+    this->text = text;
+}
+
+void ui_vertical_list::scroll_up()
+{
+    scroll = ( scroll == 0 ? text.size() - 1 : scroll - 1 );
+
+    if( scroll == text.size() - 1 ) {
+        window_scroll = scroll - get_rect().size_y + 1;
+    } else if( scroll < window_scroll ) {
+        window_scroll--;
+    }
+}
+
+void ui_vertical_list::scroll_down()
+{
+    scroll = ( scroll == text.size() - 1 ? 0 : scroll + 1 );
+
+    if( scroll > get_rect().size_y + window_scroll - 1 ) {
+        window_scroll++;
+    } else if( scroll == 0 ) {
+        window_scroll = 0;
+    }
+}
+
+const std::string &ui_vertical_list::current() const
+{
+    return text[scroll];
+}
+
+ui_horizontal_list::ui_horizontal_list( int x, int y, ui_anchor anchor ) : ui_element( 0, 1, x, y, anchor )
+{
+}
+
+void ui_horizontal_list::draw()
+{
+    auto win = get_win();
+    if( win == nullptr ) {
+        return;
+    }
+
+    unsigned int x_offset = 1 + get_ax();
+
+    for( unsigned int i = 0; i < text.size(); i++ ) {
+        x_offset += draw_subtab( win, get_ax() + x_offset, get_ay(), text[i], i == scroll ) + 2;
+    }
+}
+
+void ui_horizontal_list::set_text( std::vector<std::string> text )
+{
+    this->text = text;
+    size_t text_length = 1;
+
+    for( const auto &str : text ) {
+        text_length = utf8_width( str ) + 1;
+    }
+
+    set_rect( ui_rect( text_length, get_rect().size_y, get_rect().x, get_rect().y ) );
+}
+
+void ui_horizontal_list::scroll_left()
+{
+    scroll = ( scroll == 0 ? text.size() - 1 : scroll - 1 );
+}
+
+void ui_horizontal_list::scroll_right()
+{
+    scroll = ( scroll == text.size() - 1 ? 0 : scroll + 1 );
+}
+
+const std::string &ui_horizontal_list::current() const
+{
+    return text[scroll];
+}
+
+/////////////////////////////////////////////////////
+//THE FOLLOWING IS TEMPORARY/////////////////////////
+/////////////////////////////////////////////////////
+void label_test()
+{
+    auto lable1 = new ui_label("some", 0, 0, top_left);
```

Comment:
Some of the pointer members of the ui classes should probably be wrapped into smart pointers (e.g. `unique_ptr`) with a proper deleter function (note: the one in output.h may not be suitable as it calls `werase` and `wrefresh` before callig `delwin`).

Here I see a call to `new`, but I see no matching call to `delete` and this hints a memory leak. I assume the new label is freed from within the window that it is added to? In that case, this code makes this a bit cleared:
```C++
std::unique_ptr<ui_label> lable1( new ui_level(...) );
std::unique_ptr<ui_label> lable2( new ui_level(...) );

bordered_window win(31, 13, 50, 15);
win.add_child( std::move(lable1) ); // Ah: win owns the label now.
win.add_child( std::move(lable2) );
```

Note: adding a `make_unique` utility (function)[http://stackoverflow.com/questions/7038357/make-unique-and-perfect-forwarding] would make the first two lines a bit simple:

```C++
auto lable1 = std::make_unique<ui_level>(...);
auto lable2 = std::make_unique<ui_level>(...);
```